File: src/main/java/com/salesianostriana/dam/delight_nook/user/model/Usuario.java
package com.salesianostriana.dam.delight_nook.user.model;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.NaturalId;
import org.hibernate.proxy.HibernateProxy;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;

import java.time.Instant;
import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Entity
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@SuperBuilder
@Inheritance(strategy = InheritanceType.JOINED)
public class Usuario implements UserDetails {

    @Id
    @GeneratedValue
    @Column(columnDefinition = "uuid")
    private UUID id;

    @NaturalId
    @Column(unique = true, updatable = false)
    private String username;

    private String password;

    @Column(length = 500)
    private String nombreCompleto;

    @Column(columnDefinition = "TEXT")
    private String avatar;

    private String email;

    @ElementCollection(fetch = FetchType.EAGER)
    private Set<UserRole> roles;

    private boolean deleted;

    private boolean enabled;

    private String activationToken;

    @Builder.Default
    private Instant createdAt = Instant.now();

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return roles.stream()
                .map(role -> "ROLE_" + role)
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toSet());
    }


    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Usuario usuario = (Usuario) o;
        return getUsername() != null && Objects.equals(getUsername(), usuario.getUsername());
    }

    @Override
    public final int hashCode() {
        return Objects.hash(username);
    }
}

File: web/src/app/models/usuario.ts
import { Page } from "./page.interface";

export interface Usuario {
    username: string;
    token: string;
    refreshToken: string;
    nombreCompleto: string;
    avatar: string;
    roles: string[];
    enabled: boolean;
}

export interface UsuarioResponse {
    content: Usuario[];
    page: Page;
}

export class LoginRequest {
    private username: string;
    private password: string;

    constructor(username: string, password: string) {
        this.username = username;
        this.password = password;
    }
}

export class CreateUsuario {
    private username: string;
    private nombreCompleto: string;
    private email: string;


    constructor(username: string, nombreCompleto: string, email: string) {
        this.username = username;
        this.nombreCompleto = nombreCompleto;
        this.email = email;
    }
}

export class ValidateUsuario {
    private password: string;
    private verifyPassword: string;
    private activationToken: string;

    constructor(password: string, verifyPassword: string, activationToken: string) {
        this.password = password;
        this.verifyPassword = verifyPassword;
        this.activationToken = activationToken;
    }
}